apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+            if (_finished) {
+                submitterFinishedEvidence = _finished;
+                _challenge.submitterFinishedEvidence = _finished;
+            }
+        } else if (_submitter == _challenge.challenger) {
+            require(!challengerFinishedEvidence, ERROR_CHALLENGER_FINISHED_EVIDENCE);
+            if (_finished) {
+                submitterFinishedEvidence = _finished;
+                _challenge.challengerFinishedEvidence = _finished;
+            }
+        } else {
+            revert(ERROR_SENDER_NOT_ALLOWED);
+        }
+
+        return submitterFinishedEvidence && challengerFinishedEvidence;
+    }
+
+    /**
+    * @dev Submit evidence for an dispute on an arbitrator
+    * @param _arbitrator Arbitrator to submit evidence on
+    * @param _disputeId Identification number of the dispute on the arbitrator
+    * @param _submitter Address submitting the evidence
+    * @param _evidence Evidence data submitted for the dispute
+    * @param _finished Whether the submitter is finished submitting evidence
+    */
+    function _submitEvidence(IArbitrator _arbitrator, uint256 _disputeId, address _submitter, bytes _evidence, bool _finished) internal {
+        if (_evidence.length > 0) {
+            emit EvidenceSubmitted(_arbitrator, _disputeId, _submitter, _evidence, _finished);
+        }
+    }
+
+    /**
+    * @dev Accept a challenge proposed against an action ("reject action")
+    * @param _actionId Identification number of the action to be rejected
+    * @param _action Action instance to be rejected
+    * @param _challengeId Current challenge identification number for the action
+    * @param _challenge Current challenge instance for the action
+    */
+    function _acceptChallenge(uint256 _actionId, Action storage _action, uint256 _challengeId, Challenge storage _challenge) internal {
+        _challenge.state = ChallengeState.Accepted;
+
+        address challenger = _challenge.challenger;
+        (IDisputable disputable, CollateralRequirement storage requirement) = _getDisputableFor(_action);
+        _slashBalance(requirement.staking, _action.submitter, challenger, requirement.actionAmount);
+        _transfer(requirement.token, challenger, requirement.challengeAmount);
+        disputable.onDisputableActionRejected(_action.disputableActionId);
+        emit ActionRejected(_actionId, _challengeId);
+
+        _closeAction(_actionId, _action);
+    }
+
+    /**
+    * @dev Reject a challenge proposed against an action ("accept action")
+    * @param _actionId Identification number of the action to be accepted
+    * @param _action Action instance to be accepted
+    * @param _challengeId Current challenge identification number for the action
+    * @param _challenge Current challenge instance for the action
+    */
+    function _rejectChallenge(uint256 _actionId, Action storage _action, uint256 _challengeId, Challenge storage _challenge) internal {
+        _challenge.state = ChallengeState.Rejected;
```

Comment:
⚠️ Should we also reset the action's current challenge ID to 0 when the challenge is rejected / voided?

Reply:
Actually, thinking a bit more about this, maybe we **don't** want to reset the value.

I thought resetting it would make it easier for us to validate if an action had an ongoing challenge, but:

1. We are currently handling it fine on-chain without having to reset the value
2. Keeping the value is helpful for a frontend to easily know how many challenges an action had (its entire challenge history), without resorting to event fetching

Reply:
> Should we also reset the action's current challenge ID to 0 when the challenge is rejected / voided?

I actually tried to keep it consistent so `currentChallengeID` always references to the current/last without having to reset it. I think we are safe to keep it given the exhaustive checks we are already doing on every external function that could affect the action or challenge state. Otherwise, some of those checks are wrong and we should fix them.

I think you arrived the same conclusion reading your second comment :)